jerry-ext/jerry-module/jerry-module.c

```diff
@@ -0,0 +1,209 @@
+  memset (return_value, 0, sizeof (jerryx_module_instance_t *));
+  return return_value;
+} /* jerryx_module_context_init */
+
+void
+jerryx_module_context_deinit (void *context_p)
+{
+  if (context_p)
+  {
+    jerryx_module_link_free (*((jerryx_module_header_t **) (context_p)), jerryx_module_free_instance);
+    jmem_heap_free_block (context_p, sizeof (jerryx_module_header_t *));
+  }
+} /* jerryx_module_context_deinit */
+
+#ifdef JERRYX_MODULE_HAVE_CONTEXT
+#include "jerry-context.h"
+static int jerryx_module_user_context_slot = -1;
+JERRYX_C_CTOR (_register_module_handler_)
+{
+  jerryx_module_user_context_slot = jerryx_context_request_slot (jerryx_module_context_init,
+                                                                 jerryx_module_context_deinit);
+} /* JERRYX_C_CTOR */
+#define JERRYX_MODULE_CONTEXT \
+  (jerryx_context_get (jerryx_module_user_context_slot))
+#else /* !JERRYX_MODULE_HAVE_CONTEXT */
+#define JERRYX_MODULE_CONTEXT \
+  (jerry_get_user_context ())
+#endif /* JERRYX_MODULE_HAVE_CONTEXT */
+
+/* This list of module definitions is constructed by calls to jerry_module_register (). jerry_module_register () does
+ * not require a user context. The list is simply a collection of global static structures defined in each module,
+ * arranged into a linked list. */
+static jerryx_module_t *modules = NULL;
+
+void
+jerryx_module_register (jerryx_module_t *module_p)
+{
+  jerryx_module_link_insert ((jerryx_module_header_t *) module_p, ((jerryx_module_header_t **) (&modules)));
+} /* jerryx_module_register */
+
+void
+jerryx_module_unregister (jerryx_module_t *module_p)
+{
+  jerryx_module_link_remove (module_p->link.name, ((jerryx_module_header_t **) (&modules)));
+} /* jerryx_module_unregister */
+
+typedef struct
+{
+  jerryx_module_header_t link;
+  jerry_value_t result;
+} jerryx_instance_t;
+
+static const char not_found_prologue[] = "Module '";
+static const char not_found_epilogue[] = "' not found";
+
+jerry_value_t
+jerryx_module_load (const jerry_char_t *name)
+{
+  jerry_value_t return_value = 0;
+  size_t error_message_size = 0, name_length = strlen ((char *) name);
+  char *error_message = NULL;
+  jerryx_module_t *module = NULL;
+  jerryx_module_instance_t **instances_pp = JERRYX_MODULE_CONTEXT;
+  jerryx_instance_t *instance_p = NULL;
+
+  if (instances_pp)
+  {
+    instance_p =
+    (jerryx_instance_t *) jerryx_module_link_find (*((jerryx_module_header_t **) instances_pp), name, NULL);
+  }
+
+  if (instance_p)
+  {
+    return instance_p->result;
+  }
+
+  module = (jerryx_module_t *) jerryx_module_link_find ((jerryx_module_header_t *) modules, name, NULL);
+
+  if (module)
+  {
+    return_value = module->init ();
+    if (instances_pp)
+    {
+      instance_p = (jerryx_instance_t *) jmem_heap_alloc_block_null_on_error (sizeof (jerryx_instance_t));
+      if (instance_p)
+      {
+        JERRYX_MODULE_HEADER_RUNTIME_INIT (instance_p, module->link.name);
```

Comment:
The reason I sought to encapsulate the initialization of these linked list items is for the case that, later on, we switch to a different data structure, like a tree. Then we don't have to scour the code to make sure all the data structure -related data is up-to-date.